### PR TITLE
Publish / metadata form

### PR DESCRIPTION
**Pull request by toddmedema**
Overview:
- Removed attributes from XML, except title
- On load, if not defined, XML attributes are imported to GoogleDoc.metadata
- From there, the code pulls all metadata info from GoogleDoc.metadata
- added a parser error instructing users to delete attribute lines (see screenshot)
- validation of attributes is now in the UI / form, instead of the XML parser
- as a temporary compatibility measure, on publish, the metadata is inserted back into the XML. Once we've coded a version of the app that no longer relies on XML metadata, and we're 100% sure all users are on the new app, we can remove this stopgap code

Additional features (especially selecting genre, mood and content rating) will be rolled out separately given the large scope of this change.

screen shot 2017-05-15 at 6 24 36 pm
screen shot 2017-05-15 at 6 30 48 pm

### CODE REVIEW

File: app/actions/ActionTypes.tsx [hunk 1]
```diff
@@ -65,6 +65,12 @@ export interface QuestRenderAction extends Redux.Action {
   msgs: LogMessageMap;
 }
 
+export interface QuestMetadataChangeAction extends Redux.Action {
```
Comment (smartin015) [inline app/actions/ActionTypes.tsx]: This does limit us to single value changes. Thoughts on making this a `SetQuestMetadataAction`?

File: app/actions/quest.tsx [hunk 2]
```diff
@@ -103,71 +104,105 @@ function createDocNotes(model: any) {
   return string;
 }
 
-export function loadQuest(userid: string, dispatch: any, docid?: string) {
+function createDocMetadata(model: any, oldMeta: any) {
+  const map = model.createMap();
```
Comment (smartin015) [inline app/actions/quest.tsx]: Neat, so this should be realtime as well? Have you tried this with two different clients? What happens when:

- Client 1 publishes, with some changed values
- Client 2 clicks the publish button

Is publish state shown with the new values?

File: app/actions/quest.tsx [hunk 3]
```diff
@@ -103,71 +104,105 @@ function createDocNotes(model: any) {
...
-    getPublishedQuestMeta(userid + '_' + docid, function(quest: QuestType) {
-      var xmlResult = renderXML(text);
-      quest = Object.assign(quest || {}, xmlResult.getMeta());
-      quest.id = docid;
-      quest.mdRealtime = md;
-      quest.notesRealtime = notes;
+
+    if (!metadata) { // Create metadata if it's an old quest w/o metadata attribute
+      // Default to any metadata set in the markdown metadata (migrate from the old format)
+      const oldMeta = renderXML(md.toString()).getMeta();
```
Comment (smartin015) [inline app/actions/quest.tsx]: What if there's a render error?

File: app/actions/quest.tsx [hunk 4]
```diff
@@ -103,71 +104,105 @@ function createDocNotes(model: any) {
...
-    }
+export function publishQuestSetup(): ((dispatch: Redux.Dispatch<any>)=>any) {
+  return (dispatch: Redux.Dispatch<any>): any => {
+    dispatch({type: 'QUEST_PUBLISHING_SETUP'} as QuestPublishingSetupAction);
+  }
+}
+
+export function publishQuest(quest: QuestType): ((dispatch: Redux.Dispatch<any>)=>any) {
+  return (dispatch: Redux.Dispatch<any>): any => {
+    const xmlQdl = renderXML(quest.mdRealtime.getText());
```
Comment (smartin015) [inline app/actions/quest.tsx]: In my mind, `QDL` is the markdown language and `XML` is what it renders to. Can we rename this to `xmlRender` or some such, here and elsewhere?

File: app/actions/quest.tsx [hunk 5]
```diff
@@ -103,71 +104,105 @@ function createDocNotes(model: any) {
...
+
+    // Insert metadata back into the XML. Temporary patch until ALL client apps
+    // are running a version that supports DB metadata
```
Comment (smartin015) [inline app/actions/quest.tsx]: Add a bug reference and TODO plz.

File: app/actions/quest.tsx [hunk 6]
```diff
@@ -103,71 +104,105 @@ function createDocNotes(model: any) {
...
+        .attr('mintimeminutes', quest.mintimeminutes)
+        .attr('maxtimeminutes', quest.maxtimeminutes)
+        .toString();
+
+    dispatch({type: 'QUEST_RENDER', qdl: xmlQdl, msgs: xmlQdl.getFinalizedLogs()});
+    dispatch({type: 'REQUEST_QUEST_PUBLISH', quest} as RequestQuestPublishAction);
+    return $.ajax({
+      type: 'POST',
+      url: '/publish/' + quest.id,
+      data: xmlString,
```
Comment (smartin015) [inline app/actions/quest.tsx]: Seems like we still need to adjust the server endpoint to accept non-XML params as well, right?

File: app/components/DialogsContainer.tsx [hunk 7]
```diff
@@ -19,9 +20,30 @@ const mapStateToProps = (state: AppState, ownProps: any): DialogsStateProps => {
...
       dispatch(setDialog(dialog, false));
     },
+    onRequestPublish: (quest: QuestType): void => {
+      Joi.validate(quest, {
+        title: Joi.string().max(100),
+        summary: Joi.string().max(200),
+        email: Joi.string().email(),
+        author: Joi.string().max(100),
+        minplayers: Joi.number().min(1).max(Joi.ref('maxplayers')),
+        maxplayers: Joi.number().min(Joi.ref('minplayers')).max(6),
```
Comment (smartin015) [inline app/components/DialogsContainer.tsx]: Any way we could simply stop people from setting the values improperly in the first place? Maybe move this validation to `handleMetadataChange`?

File: app/constants.tsx [hunk 8]
```diff
@@ -8,30 +8,20 @@ export const METADATA_FIELDS = [
   'summary',
   'author',
   'email',
-  'url',
```
Comment (smartin015) [inline app/constants.tsx]: Where did URL go?

File: app/components/DialogsContainer.tsx [hunk 9]
```diff
@@ -19,9 +20,30 @@ const mapStateToProps = (state: AppState, ownProps: any): DialogsStateProps => {
...
+    handleMetadataChange: (quest: QuestType, key: string, value: any): void => {
+      dispatch(questMetadataChange(quest, key, value));
+    },
     onRequestClose: (dialog: DialogIDType): void => {
       dispatch(setDialog(dialog, false));
     },
+    onRequestPublish: (quest: QuestType): void => {
+      Joi.validate(quest, {
+        title: Joi.string().max(100),
+        summary: Joi.string().max(200),
```
Comment (smartin015) [inline app/components/DialogsContainer.tsx]: What about minimums for title, summary, author? Do these allow for empty fields?

File: app/parsing/render/BlockRenderer.tsx [hunk 10]
```diff
@@ -244,7 +244,14 @@ export class BlockRenderer {
       let k = kv[0].toLowerCase();
       let v = kv[1].trim();
       attrs[k] = v;
+
+      if (k !== 'title') {
+        if (log) {
+          log.err('Quest attributes have migrated to the "Publish" button - simply delete this line.', '420', block.startLine + i);
```
Comment (smartin015) [inline app/parsing/render/BlockRenderer.tsx]: Needs an associated markdown file. Also are we sure these values are propagated to the publish section even if they're deleted before the user hits the publish button for the first time?

File: app/components/DialogsContainer.tsx [hunk 11]
```diff
@@ -19,9 +20,30 @@ const mapStateToProps = (state: AppState, ownProps: any): DialogsStateProps => {
 
 const mapDispatchToProps = (dispatch: Redux.Dispatch<any>, ownProps: any): DialogsDispatchProps => {
   return {
+    handleMetadataChange: (quest: QuestType, key: string, value: any): void => {
+      dispatch(questMetadataChange(quest, key, value));
+    },
     onRequestClose: (dialog: DialogIDType): void => {
       dispatch(setDialog(dialog, false));
     },
+    onRequestPublish: (quest: QuestType): void => {
```
Comment (smartin015) [inline app/components/DialogsContainer.tsx]: Would be nice to see a couple unit tests to ensure that metadata changes & publishing validation are observed.

File: app/actions/ActionTypes.tsx (hunk 1 above)
Comment (toddmedema) [inline app/actions/ActionTypes.tsx]: Hmm. That'd add bloat elsewhere, as each individual form element would then need to keep track of and pass on the entire metadata state

File: app/components/DialogsContainer.tsx (hunk 7 above)
Comment (toddmedema) [inline app/components/DialogsContainer.tsx]: Correct values are mostly enforced by the UI itself (ie you can only select the numbers 1-6 for min and max players)

File: app/constants.tsx (hunk 8 above)
Comment (toddmedema) [inline app/constants.tsx]: Good point, worth a discussion. In migrating the metadata, I thought it'd make sense to remove the URL for two reasons:

1. Listing a URL in the Expedition app poses security threats - ie if users are clicking on author URLs, they're being directed to third party resources outside of our control.
2. External websites, in hindsight, seem like not a great addition to the Expedition experience. If the purpose of the URL is to provide more information about an author, I think it'd make more sense to have an account-level author bio (perhaps also moving name and email to account level) that gives authors a chance to tell users about themselves inside of the Expedition app (both a) eliminating the awkward app -> website -> app transitions and b) not requiring quest creators to host a website if they want to tell users anything about themselves).

Let me know if that makes sense / thoughts :)

File: app/actions/quest.tsx (hunk 2 above)
Comment (toddmedema) [inline app/actions/quest.tsx]: Fixed!

File: app/actions/quest.tsx (hunk 3 above)
Comment (toddmedema) [inline app/actions/quest.tsx]: What sort of render error are you thinking? Tried a couple different things like invalid metadata titles and funky indentation errors and it still caught the old metadata fine

File: app/actions/quest.tsx (hunk 4 above)
Comment (toddmedema) [inline app/actions/quest.tsx]: renderXML -> xmlRender? Not sure what the difference there is. Changed xmlQdl to questObject for a bit more clarity, since it's not actually xml being returned (you have to call `.getResult()` to actually get the XML)

File: app/actions/quest.tsx (hunk 5 above)
Comment (toddmedema) [inline app/actions/quest.tsx]: Added

File: app/actions/quest.tsx (hunk 6 above)
Comment (toddmedema) [inline app/actions/quest.tsx]: Yep, but that'll be a separate PR. Trying to keep 'em more digestible :)

File: app/components/DialogsContainer.tsx (hunk 9 above)
Comment (toddmedema) [inline app/components/DialogsContainer.tsx]: Cannot be empty - also added (very loose) minimums. Don't think we need to over-validate inputs unless we start seeing abuses :)

File: app/parsing/render/BlockRenderer.tsx (hunk 10 above)
Comment (toddmedema) [inline app/parsing/render/BlockRenderer.tsx]: Yep - migration happens on quest load :) Added an error definition

Discovered that the error documentation and testing seems to have gotten disconnected, so added an issue for that - https://github.com/ExpeditionRPG/expedition-quest-creator/issues/271

File: app/components/DialogsContainer.tsx (hunk 11 above)
Comment (toddmedema) [inline app/components/DialogsContainer.tsx]: Added some additional tests + stubs

File: app/actions/quest.tsx (hunk 3 above)
Comment (smartin015) [inline app/actions/quest.tsx]: I was thinking if something throws an exception here, the error likely wouldn't be visible to the user. And it's not unlikely that the render code would fail in a mysterious way.

File: app/actions/ActionTypes.tsx (hunk 1 above)
Comment (smartin015) [inline app/actions/ActionTypes.tsx]: After looking at the rest of the PR, this makes more sense. Let's leave it as is for now.

File: app/constants.tsx (hunk 8 above)
Comment (smartin015) [inline app/constants.tsx]: Seems fair. We'll probably end up with users posting URLs in their author bio areas... but that's probably fine for now.

File: app/actions/quest.tsx (hunk 4 above)
Comment (smartin015) [inline app/actions/quest.tsx]: My comment was on `xmlQDL` being confusing. `questObject` is less confusing, but `Object` here adds no new information. 

What about `renderResult`?

File: app/actions/quest.tsx (hunk 5 above)
Comment (smartin015) [inline app/actions/quest.tsx]: I don't see any TODO here.

File: app/components/DialogsContainer.tsx (hunk 11 above)
Comment (smartin015) [inline app/components/DialogsContainer.tsx]: We should really be testing the publish flow here, particularly validation against bad values. Can you at least add some stubs for that?

File: app/components/DialogsContainer.tsx (hunk 9 above)
Comment (smartin015) [inline app/components/DialogsContainer.tsx]: Loose minimums LGTM.

File: app/components/DialogsContainer.tsx (hunk 7 above)
Comment (smartin015) [inline app/components/DialogsContainer.tsx]: Gotcha.

File: app/components/Dialogs.tsx [hunk 12]
```diff
@@ -49,6 +52,111 @@ export class ErrorDialog extends React.Component<ErrorDialogProps, {}> {
...
+          value={metadata.get('email')}
+          floatingLabelText="Contact Email (private)"
+          onChange={(e: any, val: string) => { this.props.handleMetadataChange(this.props.quest, 'email', val)}}
+        />
+        <SelectField
+          floatingLabelText="Minimum Players"
+          value={metadata.get('minplayers')}
+          onChange={(e: any, i: number, val: number) => { this.props.handleMetadataChange(this.props.quest, 'minplayers', val)}}
+        >
+          <MenuItem value={1} primaryText="1" />
```
Comment (smartin015) [inline app/components/Dialogs.tsx]: Can we construct these based on the value in `constants.tsx` to prevent drift?

File: app/parsing/render/BlockRenderer.tsx (hunk 10 above)
Comment (smartin015) [inline app/parsing/render/BlockRenderer.tsx]: Thanks :)

File: app/components/Dialogs.tsx [hunk 13]
```diff
@@ -49,6 +52,111 @@ export class ErrorDialog extends React.Component<ErrorDialogProps, {}> {
...
+          label="Publish"
+          secondary={true}
+          onTouchTap={() => this.props.onRequestPublish(this.props.quest)}
+        />]}
+        titleClassName={'dialogTitle dialogGood'}
+        modal={false}
+        open={Boolean(this.props.open)}
+      >
+        <TextField
+          value={metadata.get('summary')}
```
Comment (smartin015) [inline app/components/Dialogs.tsx]: Min and max values based on Joi validation fields here?

File: app/actions/quest.tsx (hunk 3 above)
Comment (toddmedema) [inline app/actions/quest.tsx]: Added a try-catch

File: app/actions/quest.tsx (hunk 4 above)
Comment (toddmedema) [inline app/actions/quest.tsx]: Makes sense!

File: app/actions/quest.tsx (hunk 5 above)
Comment (toddmedema) [inline app/actions/quest.tsx]: Added a URL, but good call, added `TODO` in front

File: app/components/DialogsContainer.tsx (hunk 11 above)
Comment (toddmedema) [inline app/components/DialogsContainer.tsx]: Fair, added stubs

File: app/components/Dialogs.tsx (hunk 12 above)
Comment (toddmedema) [inline app/components/Dialogs.tsx]: Good call, added

File: app/components/Dialogs.tsx (hunk 13 above)
Comment (toddmedema) [inline app/components/Dialogs.tsx]: Material UI doesn't let you set text field max lengths in text fields, alas

Comment (smartin015) [inline app/components/Dialogs.tsx]: But it does have an `errorText` field: http://www.material-ui.com/#/components/text-field

We should be setting this (onBlur?) as the user fills out the form, instead of alerting them via popup after they try to submit. This can be a TODO.

File: app/components/Dialogs.tsx [hunk 14]
```diff
@@ -66,6 +67,14 @@ export class PublishingDialog extends React.Component<PublishingDialogProps, {}>
     if (metadata === undefined) {
       return null;
     }
+    const playerItems = [];
+    for (let i = 1; i <= MAX_PLAYERS; i++) {
+      playerItems.push(<MenuItem value={i} primaryText={i} key={i} />);
+    }
+//     (new Array(MAX_PLAYERS)).map((val: number, index: number) => {
```
Comment (smartin015) [inline app/components/Dialogs.tsx]: Remove commented code
